src/main/java/com/iota/iri/service/snapshot/impl/SnapshotProviderImpl.java

```diff
@@ -118,15 +119,38 @@ public Snapshot getLatestSnapshot() {
     }
 
     /**
-     * {@inheritDoc}
+     * {@inheritDoc}<br />
+     * <br />
+     * It first writes two temporary files and then renames them to avoid corruption of the written files in case of IRI
+     * crashes.<br />
      */
     @Override
     public void writeSnapshotToDisk(Snapshot snapshot, String basePath) throws SnapshotException {
         snapshot.lockRead();
 
+        File tmpStateFile = null;
+        File tmpMetaFile = null;
         try {
-            writeSnapshotStateToDisk(snapshot, basePath + ".snapshot.state");
-            writeSnapshotMetaDataToDisk(snapshot, basePath + ".snapshot.meta");
+            tmpStateFile = File.createTempFile("iri.", ".snapshot.state");
+            tmpMetaFile = File.createTempFile("iri.", ".snapshot.meta");
+
+            writeSnapshotStateToDisk(snapshot, tmpStateFile.getAbsolutePath());
+            writeSnapshotMetaDataToDisk(snapshot, tmpMetaFile.getAbsolutePath());
+
+            Files.move(Paths.get(tmpStateFile.getAbsolutePath()), Paths.get(basePath + ".snapshot.state"),
+                    StandardCopyOption.ATOMIC_MOVE);
+            Files.move(Paths.get(tmpMetaFile.getAbsolutePath()), Paths.get(basePath + ".snapshot.meta"),
```

Comment:
if for some reason this fails (or takes a long time) you'll be in a mid-state where the state was updated but meta not.
is this an issue? would the opposite order be better/worse? can this be done atomically as well? (move both at once).

Reply:
Usually the move should not "move" but only "rename" the file, which should be a matter of milliseconds, but I guess you are right. Especially since we create the temp files in the temp directory it could happen that the file actually gets "copied" if they reside on different file systems.

Maybe it is better to create the temp files in the same path as the target files and just append them with a .tmp extension so the rename can happen very fast.